src/peer_loop.rs

```diff
@@ -1912,33 +1914,37 @@ impl PeerLoopHandler {
                 own_datetime_utc.format("%Y-%m-%d %H:%M:%S"));
         }
 
-        // There is potential for a race-condition in the peer_map here, as we've previously
-        // counted the number of entries and checked if instance ID was already connected. But
-        // this check could have been invalidated by other tasks so we perform it again
-
-        if global_state
-            .net
-            .peer_map
-            .values()
-            .any(|pi| pi.instance_id() == self.peer_handshake_data.instance_id)
+        // There is potential for a race-condition in the peer_map here, as
+        // we've previously counted the number of entries and checked if
+        // instance ID was already connected. But this check could have been
+        // invalidated by other tasks so we perform it again here under a write
+        // lock.
```

Comment:
This comment doesn't clearly communicate why a write-lock is necessary for the check. How about:

“Up to this point, it is possible for multiple different tasks to perform connection initiation with the same node. To ensure that at most one such connection initiation attempt succeeds,
- acquire an exclusive (i.e. mutable) lock of the global state,
- assert that the node in question is not a peer, and only then
- add the node to the peer map.”

Reply:
Right. The check needs to be done while holding any lock. But since the check ends with a write to `peer_map` and we cannot release the lock before writing, the lock needs to be a write lock.

Reply:
It seems that even my suggestion above missed the point. I've made an updated suggestion inline.